server/block/wall.go

```diff
@@ -0,0 +1,244 @@
+		"wall_connection_type_north": w.NorthConnection.String(),
+		"wall_connection_type_east":  w.EastConnection.String(),
+		"wall_connection_type_south": w.SouthConnection.String(),
+		"wall_connection_type_west":  w.WestConnection.String(),
+		"wall_post_bit":              boolByte(w.Post),
+	}
+	name, meta := encodeWallBlock(w.Block)
+	if meta > 0 || name == "cobblestone" {
+		properties["wall_block_type"] = name
+		name = "cobblestone"
+	}
+	return "minecraft:" + name + "_wall", properties
+}
+
+// Model ...
+func (w Wall) Model() world.BlockModel {
+	return model.Wall{
+		NorthConnection: w.NorthConnection.String(),
+		EastConnection:  w.EastConnection.String(),
+		SouthConnection: w.SouthConnection.String(),
+		WestConnection:  w.WestConnection.String(),
+		Post:            w.Post,
+	}
+}
+
+// BreakInfo ...
+func (w Wall) BreakInfo() BreakInfo {
+	hardness := 2.0
+	name, _ := encodeWallBlock(w.Block)
+	if name == "cobbled_deepslate" || name == "deepslate_brick" || name == "deepslate_tile" || name == "polished_deepslate" {
+		hardness = 3.5
+	}
+	return newBreakInfo(hardness, pickaxeHarvestable, pickaxeEffective, oneOf(w))
+}
+
+// NeighbourUpdateTick ...
+func (w Wall) NeighbourUpdateTick(pos, _ cube.Pos, wo *world.World) {
+	w, updated := w.calculateState(wo, pos)
+	if updated {
+		wo.SetBlock(pos, w, nil)
+	}
+}
+
+// UseOnBlock ...
+func (w Wall) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, wo *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, _, used = firstReplaceable(wo, pos, face, w)
+	if !used {
+		return
+	}
+	w, _ = w.calculateState(wo, pos)
+	place(wo, pos, w, user, ctx)
+	return placed(ctx)
+}
+
+// CanDisplace ...
+func (Wall) CanDisplace(b world.Liquid) bool {
+	_, water := b.(Water)
+	return water
+}
+
+// SideClosed ...
+func (Wall) SideClosed(cube.Pos, cube.Pos, *world.World) bool {
+	return false
+}
+
+// ConnectionType returns the connection type of the wall in the given direction.
+func (w Wall) ConnectionType(direction cube.Direction) WallConnectionType {
+	switch direction {
+	case cube.North:
+		return w.NorthConnection
+	case cube.East:
+		return w.EastConnection
+	case cube.South:
+		return w.SouthConnection
+	case cube.West:
+		return w.WestConnection
+	}
+	panic("unknown direction")
+}
+
+// WithConnectionType returns the wall with the given connection type in the given direction.
+func (w Wall) WithConnectionType(direction cube.Direction, connection WallConnectionType) Wall {
+	switch direction {
+	case cube.North:
+		w.NorthConnection = connection
+	case cube.East:
+		w.EastConnection = connection
+	case cube.South:
+		w.SouthConnection = connection
+	case cube.West:
+		w.WestConnection = connection
+	}
+	return w
+}
+
+// calculateState returns the wall with the correct state based on walls around it. If any of the connections have been
+// updated then the wall and true are returned.
+func (w Wall) calculateState(wo *world.World, pos cube.Pos) (Wall, bool) {
```

Comment:
Might be helpful to add some documentation here and space it out a bit more so it's more readable.